### DISCUSSION TITLE
Communication between App Services

### DISCUSSION THREAD

**Discussion by vijay-hcl-epl** [edgexfoundry/community#196] [Q&A] [answered]
Hi EdgeX Team, 
Currently we are exploring all the EdgeX features for one of our products. I need to know how far the communication between application services are possible. I read from this link - [https://docs.edgexfoundry.org/3.0/microservices/application/ApplicationServices/](url) that 

> Application Services provide the means for data to be prepared (transformed, enriched, filtered, etc.) and groomed (formatted, compressed, encrypted, etc.) before being sent to an endpoint of choice or **published back to other Application Service to consume.**

I would like to understand more about "**_published back to other Application Service to consume._**"

1. Can one application service directly publish the data to other application service? I hope it is not possible and we need to use "_SetResponseData_" so that the data goes to message bus and then to the other application service. Please confirm.
2. If I use an app service with MQTT/HTTP/Custom trigger, then is it possible to publish the data from that service to another app service which uses Message bus trigger?
3. If I use two app services, one with HTTP trigger and another with MQTT trigger, then is it possible to share data between them?
4. Can I POST an event to core data using the api "_/event/{profileName}/{deviceName}/{sourceName}_" either from an application service or any external service, so that another application service with message bus trigger can read this POSTed event?

It will be helpful if you give a inline response.

Thanks in advance,
Vijay

**Comment by lenny-goodell** [comment, marked as answer]
@vijay-hcl-epl , you can send data to other App service in multiple ways. #1 or #2 are the prefered was. Choosing which one depends on how your data it processed. If fully process in the function pipeline, then chose #1. If combined and processed in background then choose #2.

1. Use SetResponseData to publish the data to the MessageBus using a specific topic (from config) which the other App Service is subscribed. If your data has been transformed to a custom type the other App Service must know that type and set TargetType.
https://docs.edgexfoundry.org/3.1/microservices/application/sdk/api/AppFunctionContextAPI/#setresponsedata
https://docs.edgexfoundry.org/3.1/microservices/application/sdk/details/TargetType/
2. Similar to above and new in EdgeX 3.1, you can use the Publish/PublishWithTopic functions to Publish the data to the MessageBus.
https://docs.edgexfoundry.org/3.1/microservices/application/sdk/api/ApplicationServiceAPI/#publishwithtopic
https://docs.edgexfoundry.org/3.1/microservices/application/sdk/api/AppFunctionContextAPI/#publishwithtopic
3. Export the data via Http Export function the the Http trigger of the other App Service. Same consideration for custom type as above.
https://docs.edgexfoundry.org/3.1/microservices/application/details/Triggers/
https://docs.edgexfoundry.org/3.1/microservices/application/sdk/api/BuiltInPipelineFunctions/#http-export
4. Use custom REST endpoints to communicate between the App Services.
https://docs.edgexfoundry.org/3.1/microservices/application/sdk/details/CustomRestApis/

**Reply by vijay-hcl-epl** [reply]
Thank you @lenny-intel . This helps..!!
